src/wpilib/i2c.rs

```diff
@@ -0,0 +1,87 @@
+use wpilib::wpilib_hal::*;
+
+/// Helper enum that describes one of two possible locations on the roborio for i2c devices
+#[derive(Copy, Clone)]
+pub enum I2cPort {
+    /// Enum option for i2c device plugged into the onboard i2c connectors on the roborio
+    OnBoard = 0,
+    /// Enum option for i2c device, such as a navX board, connected through the MXP port
+    MXP = 1,
+}
+
+/// Describes various errored or otherwise special cases that may result from an i2c transaction
+pub enum I2cError {
+    // TODO find is there is any way to get more i2c transaction information out of HAl than just
+    // success or failure
+
+    /// Indicates a general error in an i2c transaction
+    IOError,
+    /// Indicates when no string, or an invalid string was received from the i2c transaction
+    InvalidReceiveString,
+    /// Indicates a situation where the connection was specifically aborted
+    TransferAbort,
+}
+
+/// Struct for sending and receiving data over i2c
+pub struct I2cInterface {
+    port: I2cPort,
+    device_address: i32,
+}
+
+impl I2cInterface {
+    /// users create a new I2cInterface here
+    pub fn new(p: I2cPort, addr: i32, init_status: &mut i32) -> I2cInterface {
+        unsafe { HAL_InitializeI2C(p as i32, init_status); };
+        I2cInterface { port: p, device_address: addr }
+    }
+
+    /// perform a simultaneous read from and write to an i2c device
+    pub fn transaction(&mut self, sent: &mut [u8], received: &mut [u8]) -> Result<(), I2cError> {
```

Comment:
Something I missed again in my review: `sent` really doesn't have to be `mut`, as `HAL_TransactionI2C` takes a `mut` pointer only because it's autogenerated from C. The actual function does not mutate the value, and in fact casts the pointer to a `const char*` right away. You should be able to get away with just casting `sent` to a mutable pointer in the unsafe block. It isn't necessarily a "clean" solution, but neither is this function taking in a mutable slice when it doesn't need mutability on the sent data.